fuel-txpool/src/containers/dependency.rs

```diff
@@ -0,0 +1,632 @@
+                                gas_price: GasPrice::MAX,
+                            })
+                            .used_by
+                            .insert(tx.id());
+                    }
+
+                    // yey we got our contract
+                }
+            }
+        }
+        Ok((max_depth, db_coins, db_contracts, collided))
+    }
+    /// insert tx inside dependency
+    /// return list of transactions that are removed from txpool
+    pub async fn insert<'a>(
+        &'a mut self,
+        txs: &'a HashMap<TxId, ArcTx>,
+        db: &dyn TxPoolDB,
+        tx: &'a ArcTx,
+    ) -> anyhow::Result<Vec<ArcTx>> {
+        let (max_depth, db_coins, db_contracts, collided) = self.check_for_colision(txs, db, tx)?;
+
+        // now we are sure that transaction can be included. remove all collided transactions
+        let mut removed_tx = Vec::new();
+        for collided in collided.into_iter() {
+            let collided = txs
+                .get(&collided)
+                .expect("Collided should be present in txpool");
+            removed_tx.extend(self.recursively_remove_all_dependencies(txs, collided.clone()));
+        }
+
+        // iterate over all inputs and spend parent coins/contracts
+        for input in tx.inputs() {
+            match input {
+                Input::Coin { utxo_id, .. } => {
+                    // spend coin
+                    if let Some(state) = self.coins.get_mut(utxo_id) {
+                        state.is_spend_by = Some(tx.id());
+                    }
+                }
+                Input::Contract { contract_id, .. } => {
+                    // contract
+                    if let Some(state) = self.contracts.get_mut(contract_id) {
+                        state.used_by.insert(tx.id());
+                    }
+                }
+            }
+        }
+
+        // insert all coins/contracts that we got from db;
+        self.coins.extend(db_coins.into_iter());
+        // for contracts from db that are not found in dependency, we already inserted used_by
+        // and are okay to just extend current list
+        self.contracts.extend(db_contracts.into_iter());
+
+        // iterate over all outputs and insert them, marking them as available.
+        for (index, output) in tx.outputs().iter().enumerate() {
+            let utxo_id = UtxoId::new(tx.id(), index as u8);
+            match output {
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
+                    // insert output coin inside by_coin
+                    self.coins.insert(
+                        utxo_id,
+                        CoinState {
+                            is_spend_by: None,
+                            depth: max_depth,
+                        },
+                    );
+                }
+                Output::ContractCreated { contract_id } => {
+                    // insert contract
+                    self.contracts.insert(
```

Comment:
What should we do if two txs that don't collide contain the same output contract id?

Reply:
We should probably reject the later one unless it does RBF (iirc, pays more than some multiplier more fees)

Reply:
Very unlikely to happen, we are talking here about sha3 colision: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/identifiers.md#contract-id

Reply:
@Voxelot are you referring to the _contract ID_, or the ID of the TXO of the contract? The latter would require a collision in the contract ID in the first place.

Reply:
Wait, witness in contractid is not a signature, but a bytecode? That means someone can construct same contractid, and this collision check should be needed?

Reply:
The only time witness data is not a signature in a way that's interpreted by the consensus rules is for contract creation, one of the witnesses will be contract bytecode (but the consensus rules check that the contract ID of _that_ bytecode must match up with the contract ID in the contract created output).

Someone could try to create a contract with the same contract ID, but there must be a check that prevents this (cannot have two contract UTXOs with the same contract ID).

Reply:
ContractId collision check added.